synth_node_bin/src/action/advanced_sn_for_s001.rs

```diff
@@ -0,0 +1,181 @@
+                },
+                Ok((src, msg)) = synth_node.try_recv_message() => {
+                    handle_rx_msg(synth_node, src, msg).await?;
+                },
+            }
+        }
+    }
+}
+
+async fn trace_debug_info(synth_node: &SyntheticNode) {
+    let peer_infos = synth_node.connected_peer_infos();
+    let peer_cnt = peer_infos.len();
+
+    let mut log = format!("\nNumber of peers: {peer_cnt}\n");
+
+    // Let's sort by the connection's time value.
+    let mut peer_infos: Vec<_> = peer_infos.iter().collect();
+    peer_infos.sort_by(|a, b| a.1.stats().created().cmp(&b.1.stats().created()));
+
+    for (addr, info) in peer_infos.iter() {
+        let stats = info.stats();
+
+        // Align all possible IP addresses (both v4 and v6) vertically
+        const MAX_IPV6_ADDR_LEN: usize = 37;
+
+        log.push_str(&format!(
+            "{addr:>ident$}   ({side:?}) seconds: {time:?}\n",
+            addr = addr,
+            ident = MAX_IPV6_ADDR_LEN,
+            side = info.side(),
+            time = stats.created().elapsed()
+        ));
+    }
+
+    tracing::info!("{log}");
+}
+
+fn broadcast_periodic_msgs(synth_node: &mut SyntheticNode) -> Result<()> {
+    if let Err(e) = broadcast_ping_msg(synth_node) {
+        tracing::warn!("failed to broadcast Ping messages: {e}");
+    } else if let Err(e) = broadcast_addr_msg(synth_node) {
+        tracing::warn!("failed to broadcast Addr messages: {e}");
+    } else if let Err(e) = broadcast_get_addr_msg(synth_node) {
+        tracing::warn!("failed to broadcast GetAddr messages: {e}");
+    }
+
+    Ok(())
+}
+
+fn broadcast_ping_msg(synth_node: &mut SyntheticNode) -> Result<()> {
+    let msg = Message::Ping(Nonce::default());
+
+    for addr in synth_node.connected_peers() {
+        if synth_node.unicast(addr, msg.clone()).is_err() {
+            tracing::error!("failed to send {msg:?} to {addr}\n");
+            anyhow::bail!("connection closed");
+        }
+    }
+
+    Ok(())
+}
+
+fn broadcast_get_addr_msg(synth_node: &mut SyntheticNode) -> Result<()> {
+    let msg = Message::GetAddr;
+
+    for addr in synth_node.connected_peers() {
+        if synth_node.unicast(addr, msg.clone()).is_err() {
+            tracing::error!("failed to send {msg:?} to {addr}\n");
+            anyhow::bail!("connection closed");
+        }
+    }
+
+    Ok(())
+}
+
+async fn handle_rx_msg(
+    synth_node: &mut SyntheticNode,
+    src: SocketAddr,
+    msg: Message,
+) -> Result<()> {
+    tracing::info!("message received from {src}:\n{msg:?}");
+
+    // We are only handling a GetAddr for now.
+    if msg == Message::GetAddr {
+        unicast_addr_msg(synth_node, src)?;
+    }
+
+    Ok(())
+}
+
+fn unicast_addr_msg(synth_node: &mut SyntheticNode, dst: SocketAddr) -> Result<()> {
+    // These are all our IPs
+    let addrs = vec![
+        // Our zebrad
+        NetworkAddr::new(SocketAddr::from_str("35.210.208.185:8233").unwrap()),
```

Comment:
It could be configurable in the future to easily change advertised addresses. It could be useful to switch between testnet and mainnet addresses.
TBH I find it could be more useful to change those list from configuration file than setting a listening IP address (but it's only my feeling) 😃

Reply:
for now it's hardcoded, I didn't wanna fit too much stuff in this PR - will be more flexible in the future

Reply:
Great to hear! 👍